website/content/en/docs/advanced-topics/heterogeneous.md

```diff
@@ -0,0 +1,400 @@
+
+### Determining the Supported Architectures for an Operator/Operand Image
+For a given container image, you can check which platforms are supported by listing them by inspecting the manifest.
+```bash
+docker manifest inspect <image>
+```
+If the above command fails, it may be because the referenced name points to a single image rather than a manifest list. In this case, you can find the architecture by inspecting the image.
+```bash
+docker inspect <image>
+```
+
+Here's an example of the architectures distributed for the Alpine Linux container image.
+```bash
+$ docker manifest inspect alpine | grep architecture
+            "architecture": "amd64",
+            "architecture": "arm",
+            "architecture": "arm",
+            "architecture": "arm64",
+            "architecture": "386",
+            "architecture": "ppc64le",
+            "architecture": "s390x",
+```
+
+#### Shipping a Multi-Architecture Operator
+If once you've determined which subset of architecture nodes you want to distribute your operator for, there are tools available to help you distribute your operator.
+
+OperatorSDK provides a validator to help check bundle configuration.
+```bash
+operator-sdk bundle validate ./bundle --select-optional name=multiarch
+```
+
+If you're planning to distribute the operator via OLM, you can find more information in the [OLM guide for supporting multiarch](https://olm.operatorframework.io/docs/advanced-tasks/ship-operator-supporting-multiarch/).
+
+### Setting the Supported Architectures Operator & Operand Workloads
+[Kubernetes provides a mechanism](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/) called `nodeAffinity` which can be used to limit the possible node targets where a pod can be scheduled. The following example can be used to update a `PodSpec` or `PodTemplateSpec` to prevent the scheduling of pods on nodes of incompatible architectures. Here we compare the `kubernetes.io/arch` key set on the node to ensure that the value matches one of the supported architectures. This assumes that the referenced image points to a manifest list of images for each of the architectures supported. It is important to remember that `nodeAffinity` must be set anywhere a container reference is defined, including `Pod`, `Deployment`, `DaemonSet`, `StatefulSet`, or any other object that defines a `PodSpec` or `PodTemplateSpec`.
+
+```yaml
+affinity:
+  nodeAffinity:
+    requiredDuringSchedulingIgnoredDuringExecution:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: kubernetes.io/arch
+          operator: In
+          values:
+          - amd64
+          - arm64
+          - ppc64le
+          - s390x
+```
```

Comment:
can we also add the "operating system" piece in the `matchExpressions` in the code snippet too?

We've started to get asks around how to specify the workloads managed by Operators to be scheduled to the linux node pool in a Windows container/Windows node enabled mixed cluster. Hence it'll be good to have a single doc that covers how to specify arch and OS in nodeAffinity so folks can follow.


- matchExpressions:
          - key: kubernetes.io/os
            operator: In
            values:
              - linux

Reply:
@tlwu2013 pretty sure that example boils down to `(arch == amd64 || arch == arm64 || arch == ppc64le || arch == s390x) || (os == linux)`, which doesn't seem like what you want?

Seems like you'd actually want:
```yaml
affinity:
  nodeAffinity:
    requiredDuringSchedulingIgnoredDuringExecution:
      nodeSelectorTerms:
      - matchExpressions:
        - key: kubernetes.io/arch
          operator: In
          values:
          - amd64
          - arm64
          - ppc64le
          - s390x
        - key: kubernetes.io/os
          operator: In
          values:
            - linux   
```

Reply:
I am also worried about [OLM's CSV multi-arch annotations](https://olm.operatorframework.io/docs/advanced-tasks/ship-operator-supporting-multiarch/#supported-labels) which are ambiguous if multiple architectures and OSes are supported.

If someone had an operator that supported windows/amd64 and linux/arm64, there's no way to precisely assert that with OLM's annotations. You'd probably end up putting:
```
labels:
    operatorframework.io/arch.amd64: supported
    operatorframework.io/arch.arm64: supported
    operatorframework.io/os.linux: supported
    operatorframework.io/os.windows: supported
```

Reply:
Seems like we need to deprecate those labels and replace them with something like:
```yaml
labels:
    operatorframework.io/platform.windows_amd64: supported
    operatorframework.io/platform.linux_arm64: supported
```

Reply:
> @tlwu2013 pretty sure that example boils down to `(arch == amd64 || arch == arm64 || arch == ppc64le || arch == s390x) || (os == linux)`, which doesn't seem like what you want?
> 
> Seems like you'd actually want:
> 
> ```yaml
> affinity:
>   nodeAffinity:
>     requiredDuringSchedulingIgnoredDuringExecution:
>       nodeSelectorTerms:
>       - matchExpressions:
>         - key: kubernetes.io/arch
>           operator: In
>           values:
>           - amd64
>           - arm64
>           - ppc64le
>           - s390x
>         - key: kubernetes.io/os
>           operator: In
>           values:
>             - linux   
> ```

good catch @joelanford, the example I pasted was copied from https://github.com/kubernetes-sigs/kubebuilder/pull/2906. So sounds like that might need to be changed too.